### DISCUSSION TITLE
Are forms parsed when called from controller?

### DISCUSSION THREAD

**Discussion by colmexdev** [jeremyevans/forme#129] [General]
Hi there!

Silly question: Does Forme serialize the generated form when called from inside route's code (e.g. when used for `Sequel::Model` attributes' setting)? If so, is there a way to avoid penalty for view serialization?

Thanks in advance!

**Comment by jeremyevans** [comment]
Are you asking about the `forme_set` Roda plugin, or something else?  Could you provide some example code so I can understand the context?

**Reply by colmexdev** [reply]
Sorry for lack of clarity. Here's a brief example:

Suppose I have a model `Model1 < Sequel::Model` and in my routing tree I have the following:

```ruby
class App < Roda
  route do |r|
    # Maybe optimizing a bit could be helpful.
    r.is "new" do
      @model = Model1.new
      r.get do
        view("new")
      end
      r.post do
        model_form(@model)
        model.forme_set(r.params["model"]) # Here I call the Forme's form not in view, where I'm curious to know if it parses the whole form as HTML
        if @model.valid?
          @model.save
          r.redirect "/show"
        else
          view("new")
        end
      end
    end
    # Other routes for completeness.
  end
end
```

And I have this method in a helper:
```ruby
def model_form(model)
  # As you can see, the form includes HTML attributes and classes to consider on view rendering.
  form(model, { action: ENV.fetch("ACTION"), class: "flex gap-x-6 gap-y-8", enctype: "multipart/form-data" }, emit: false, labeler: :explicit, wrapper: :div) do |f|
    f.input(:attr1, class: "custom-input", wrapper_attr: { class: "custom-wrapper" }, label_attr: { class: "custom-label" }, error_attr: { class: "label-styles" }, label: "Attribute 1")
    f.input(:attr2, class: "custom-input", wrapper_attr: { class: "custom-wrapper" }, label_attr: { class: "custom-label" }, error_attr: { class: "label-styles" }, label: "Attribute 2")
    f.input(:attr3, class: "custom-input", wrapper_attr: { class: "custom-wrapper" }, label_attr: { class: "custom-label" }, error_attr: { class: "label-styles" }, label: "Attribute 3")
    f.input(:attr4, class: "custom-input", wrapper_attr: { class: "custom-wrapper" }, label_attr: { class: "custom-label" }, error_attr: { class: "label-styles" }, label: "Attribute 4")
    # And so on...
  end
end
```

I call the helper method on view like this:
```html
<%== model_form(@model) %>
```

And it parses correctly. But as you stated in a previous discussion, Forme can slow down a bit on parsing and serializing to HTML code, so that's why I've been wondering if form's serialization applies exclusively on view rendering or it also goes on while calling the helper method on route's code.

Hope that helps.

**Reply by jeremyevans** [reply]
If you are using Roda, I recommend using the `forme_set` Roda plugin, in addition to the `forme_set` model plugin.  Then you can do:

```ruby
  r.post do
    forme_set(@model)
    if @model.valid?
    # ...
```

**Reply by colmexdev** [reply]
Ok, this works great! Thank you for your suggestion.

OTOH, would you recommend to have 2 methods for the same form, one for HTML parsing on view and one for route code (without HTML attributes sepcified) or it doesn't parse form as HTML on route code?

**Reply by jeremyevans** [reply]
With the Roda `forme_set` plugin, the form submission includes signed metadata describing the contents of the form, so you don't need to render the form inside the submission handler.

**Reply by colmexdev** [reply]
I get it now, One last thing then: Would you recommend to use the session's secret into the Roda's `forme_set` plugin? Or would it be better for it to be a different one?

**Reply by jeremyevans** [reply]
You can keep it the same, but security wise it is probably best if it is different. I often use a primary secret, and have the session_secret, forme_set secret, and other secrets be generated from the primary secret using an HMAC.

**Reply by colmexdev** [reply]
Thank you very much for your advice @jeremyevans ! 😀
